# Notebook 2: Exploring the Data I


## 2.1: Seeing the Problem in Data

Finding the origins of cholera's spread was a contentious issue in the 1800s. Before even seeing the problem in the data, the people of that time could see cholera **all around them**. As friends and relatives grew gravely ill, it became urgent to discover **why** this was happening in hopes of putting a stop to it. 

People, including the local media, had different ideas about what could be causing cholera. For instance, take a look at the following political cartoon of the time: 
<br>

<table><tr>
    <td> <img src="imgs/king_cholera.png" alt="Drawing" style="width: 400px;"/> </td>
</tr></table>

<br>

<img src="imgs/pencil.png" alt="Drawing" align=left style="width: 20px;"/> <font size=4> **Journal 2a:** Interpret the Cartoon </font>

**What is the underlying message of this cartoon?** 

> Write your answer here! 


### Data Science 101: Finding the Problem

Data scientists have curious minds; when confronted with a problem in the 'real world' they first try to better understand that problem with data (before, of course, looking for answers and solutions). 

After all, it's a lot easier for superheroes to solve a mystery when there's a signal illuminating what is driving the problem. 

<table><tr>
    <td> <img src="imgs/bat-signal.jpeg" alt="Drawing" style="width: 400px;"/> </td>
</tr></table>

<br><br>

Using the Pandas toolkit, let's see if we can dig into our data to gain any insight as to ***when*** cholera outbreaks have occurred. The data below show, for a given year, how many people lived in London and how many people died there (of any cause).

In [1]:
# Load our Pandas data science library
import pandas as pd

In [2]:
# Load data about London
London = pd.read_csv("Datasets/London.csv", index_col='Year')
London

Population  Deaths
Year                    
1840     1842458   46281
1841     1877963   45284
1842     1916860   45272
1843     1953787   48574
1844     2033816   50423
1845     2073298   48332
1846     2113535   49089
1847     2195401   60442
1848     2238703   57628
1849     2282858   68432
1850     2327884   48579
1851     2373799   55354
1852     2420619   54213
1853     2468362   61202
1854     2517048   73697

<img src="imgs/pencil.png" alt="Drawing" align=left style="width: 20px;"/> <font size=4> **Journal 2b:** Thinking about the Data </font>

**In 1-3 sentences, comment on what you see in these data. Is there enough here to determine when cholera outbreaks occurred? Why or why not?** 

> Write your answer here! 

### The importance of normalization 

'Deaths' are higher in 1854 than in 1840. Is this because of cholera? *Maybe*. Is this because of population growth? *Possibly*. Simply put, if you have more people in a city, then more people die. 

For instance, if `40,000` people die in Chicago each year (pop. 3,000,000), and only `50` people die in the small farm town of Lonsdale, MN each year (pop. 4,000) ... then are you `40,000 / 50 = 800` times less likely to die in Lonsdale?!


<table><tr>
    <td> <img src="imgs/chicago.jpeg" alt="Drawing" style="width: 400px;"/> </td>
    <td> <img src="imgs/lonsdale.jpeg" alt="Drawing" style="width: 400px;"/> </td>
</tr></table>

This example highlights the importance of ***normalization***: adjusting the values of data so that they are on the **same scale**  – in this case, so you can compare the chance of dying in the much larger city of Chicago vs. the much smaller town of Lonsdale. 


<br><br>
<img src="imgs/pencil.png" alt="Drawing" align=left style="width: 20px;"/> <font size=4> **Journal 2c**: How to Normalize
    
**For the London data, how might you normalize to find years when cholera is particularly fatal?**</font>

> Write your answer here! 

<br><br>

## Creating an Outcome Variable
An **outcome variable** is the variable that we want to explain using other variables! You can **normalize an outcome variable** to avoid the population-level pitfalls discussed earlier. 

Let's return to the London example... 

Since we are interested in reasons why people die of cholera, `Deaths` seems like a logical choice for our outcome variable! 

BUT different years have different populations, which leads us to the "NYC-vs-Lonsdale" dilemma from before... 

We can easily normalize `Deaths` by using the `Population` variable to create a `Death Rate`, also known as "Mortality Rate".

This is done with the following calculation:
$$death \ rate_{1000} = {deaths \over population} \times 1000$$

In [10]:
# Calculate mortality rate
London['Deaths'] / London['Population'] * 1000

# Calculate mortality rate per 100k. 
# The "/" means that we divide every item in the "Deaths" column by every item in the "Population" column.
# The "*" means that we multiply every value (for our new Outcome variable) by 1000. 
London['Deaths per 1000'] = London['Deaths'] / London['Population'] * 1000

London

Population  Deaths  Deaths per 1000
Year                                     
1840     1842458   46281        25.119161
1841     1877963   45284        24.113361
1842     1916860   45272        23.617792
1843     1953787   48574        24.861461
1844     2033816   50423        24.792312
1845     2073298   48332        23.311651
1846     2113535   49089        23.226017
1847     2195401   60442        27.531189
1848     2238703   57628        25.741691
1849     2282858   68432        29.976459
1850     2327884   48579        20.868308
1851     2373799   55354        23.318739
1852     2420619   54213        22.396337
1853     2468362   61202        24.794580
1854     2517048   73697        29.279140

<img src="imgs/pencil.png" alt="Drawing" align=left style="width: 20px;"/> <font size=4> **Journal 2c**: How to Normalize
    
**Explain in your own words why `Deaths per 1000` is a better outcome variable than `Deaths`?**</font>

> Write your answer here! 


<img src="imgs/pencil.png" alt="Drawing" align=left style="width: 20px;"/> <font size=4> **Journal 2d**: How to Normalize
    
**What years, based on your new outcome variable, are there cholera outbreaks?**</font>

> Write your answer here! 

## 2.2: Pandas Dataframe Manipulations

-------------------------


In the previous Pandas dataframe, we created a new column by combining the other columns in some meaningful way. In this section, we will practice dataframe manipulations. To begin, let's create a small dataframe containing synthetic data as follows --- "A large rainstorm hit the imaginary country of Datapan. The following dataset contains city names, city populations, and accumulated rainfall (in inches)": 

In [16]:
column_labels = ['city', 'rainfall_in', 'population']
location_1 = ['alicetown', 17.0, 355]
location_2 = ['bobburg', 45.5, 1001]
location_3 = ['charlie city', 1.5, 845]

locations = [location_1, location_2, location_3]
rain_df = pd.DataFrame(locations, columns=column_labels)
rain_df


city  rainfall_in  population
0     alicetown         17.0         355
1       bobburg         45.5        1001
2  charlie city          1.5         845

### Action 1. Reading columns.
-----------------------------------
Let's say that you want to 'select' the values in some number of dataframe columns. You can select just one column using `df["column_name"]` or multiple columns as follows `df[["column_name_1", "column_name_2"]]`. See the following...

In [35]:
# Select only the rainfall_in column. 
new_df = rain_df["rainfall_in"]
new_df

0    17.0
1    45.5
2     1.5
Name: rainfall_in, dtype: float64

In [20]:
# Select both the city and population columns. 
rain_df[["city", "population"]]

city  population
0     alicetown         355
1       bobburg        1001
2  charlie city         845

### Action 2. Creating a new column. 
------------------------------------------
Let's say that you want to create a new column in Pandas. This can be done by setting a column to a value by using the `df["column_name"] = ...` notation. Let's create some new columns, as follows: 

In [22]:
# A. Create a new column that adds 7 and 10. 
rain_df['7_plus_10'] = 7 + 10
rain_df

city  rainfall_in  population  7_plus_10
0     alicetown         17.0         355         17
1       bobburg         45.5        1001         17
2  charlie city          1.5         845         17

In [26]:
# B. Create a new column that counts how many 'hundreds' of people there are.
rain_df['population_hundreds'] = rain_df["population"] / 100
rain_df

city  rainfall_in  population  7_plus_10  population_hundreds
0     alicetown         17.0         355         17                 3.55
1       bobburg         45.5        1001         17                10.01
2  charlie city          1.5         845         17                 8.45

In [29]:
# C. Create a new column that subtracts the rainfall from the population. 
rain_df['pop_minus_rain'] = rain_df['population'] - rain_df['rainfall_in']
rain_df  # The pop_minus_rain column might be useless, but we sure are getting good at Pandas! 

city  rainfall_in  population  7_plus_10  population_hundreds  \
0     alicetown         17.0         355         17                 3.55   
1       bobburg         45.5        1001         17                10.01   
2  charlie city          1.5         845         17                 8.45   

   pop_minus_rain  
0           338.0  
1           955.5  
2           843.5

### Action 3. Column statistics. 
-------------------------------------
As data scientists, we often want to know *something* about entire columns of data. 

Pandas provides a number of utilities to compute column **aggregates**, as follows: 

- **sum**: add up elements in a column. 
- **mean**: compute the average of the column. 
- **min**: find the minimum value(s) of a column. 
- **max**: find the maximum value(s) of a column. 

In [42]:
rain_min = rain_df['rainfall_in'].min()
rain_max = rain_df['rainfall_in'].max()
rain_mean = rain_df['rainfall_in'].mean()
rain_sum = rain_df['rainfall_in'].sum()

print(f"Rainfall min: {rain_min}\nRainfall max: {rain_max}\nRainfall Mean: {rain_mean}\nRainfall Sum: {rain_sum}")

Rainfall min: 1.5
Rainfall max: 45.5
Rainfall Mean: 21.333333333333332
Rainfall Sum: 64.0


### Coding Exercise
Perform the following dataframe manipulation exercises! Fill in the "???" with the proper Python code!

In [49]:
# Before, we begin, let us clear out our 'practice columns'.
rain_df = rain_df.drop(['7_plus_10', 'population_hundreds', 'pop_minus_rain'], axis=1, errors='ignore')
rain_df

city  rainfall_in  population
0     alicetown         17.0         355
1       bobburg         45.5        1001
2  charlie city          1.5         845

***1. Print out a version of `rain_df` containing only city names and populations. Call this dataframe `city_info_df`.***

In [ ]:
# Put your answer here!
city_info_df = rain_df???

***2. Create a column called rainfall_per_capita that is the total rainfall per person *hint: see Outcome variable*.***

In [ ]:
# Put your answer here! 
rainfall_df[???] = rainfall_df??? ??? rainfall_df["population"]

***3. A new meteorologist comes to town and states that every rain gauge in Datapan is reading two inches short! Add two inches to every `rainfall_in` entry in `rain_df` *hint: you can overwrite a column by writing to a new column of the same name!***

In [ ]:
rainfall_df[???] = ??? 

### 2.3 Reflection

<img src="imgs/pencil.png" alt="Drawing" align=left style="width: 20px;"/> <font size=4> **Journal 2e**: The cartoon
    
**See the cartoon at the start of this notebook. What variables would you want to use to \*explore or explain\* your outcome variable?** Why? Come up with a potential explanation of how elevation or home value, etc., could impact the spread of cholera?</font>

> Write your answer here! 

<img src="imgs/pencil.png" alt="Drawing" align=left style="width: 20px;"/> <font size="4">**Journal 2f:** Wrap up </font>

**What did you learn in this notebook?**
> Write your answer here! 

**What would you like to learn in the upcoming Data4All lessons?**
> Write your answer here! 